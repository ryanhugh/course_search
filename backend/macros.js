/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import path from 'path';
import URI from 'urijs';
import fs from 'fs-promise';
import rollbar from 'rollbar';
import Amplitude from 'amplitude';

import commonMacros from '../common/abstractMacros';

const amplitude = new Amplitude(commonMacros.amplitudeToken);

// Collection of small functions that are used in many different places in the backend.
// This includes things related to saving and loading the dev data, parsing specific fields from pages and more.
// Would be ok with splitting up this file into separate files (eg, one for stuff related to scraping and another one for other stuff) if this file gets too big.
// Stuff in this file can be specific to the backend and will only be ran in the backend.
// If it needs to be ran in both the backend and the frontend, move it to the common macros file :P

// TODO: improve getBaseHost by using a list of top level domains. (public on the internet)

// Change the current working directory to the directory with package.json and .git folder.
const originalCwd = process.cwd();
let oldcwd;
while (1) {
  try {
    fs.statSync('package.json');
  } catch (e) {
    oldcwd = process.cwd();
    //cd .. until in the same dir as package.json, the root of the project
    process.chdir('..');

    // Prevent an infinate loop: If we keep cd'ing upward and we hit the root dir and still haven't found
    // a package.json, just return to the original directory and break out of this loop.
    if (oldcwd === process.cwd()) {
      console.log("Can't find directory with package.json, returning to", originalCwd);
      process.chdir(originalCwd);
      break;
    }

    continue;
  }
  break;
}


// This is the JSON object saved in /etc/searchneu/config.json
// opened once getEnvVariable is called once.
let envVariablesPromise = null;

class Macros extends commonMacros {
  static parseNameWithSpaces(name) {
    // Standardize spaces.
    name = name.replace(/\s+/gi, ' ');

    // Generate first name and last name
    const spaceCount = Macros.occurrences(name, ' ', false);
    const splitName = name.split(' ');


    if (spaceCount === 0) {
      Macros.critical('0 spaces found in name', name);
      return null;
    }

    // Handles firstName, lastName and firstName, middleName, lastName

    if (spaceCount > 2) {
      Macros.log(name, 'has more than 1 space in their name. Using first and last word.');
    }

    const obj = {};

    obj.firstName = splitName[0];
    obj.lastName = splitName[splitName.length - 1];

    return obj;
  }

  // Standardizes email addresses found across different pages
  // Removes a 'mailto:' from the beginning
  // Ensures the email contains a @
  static standardizeEmail(email) {
    if (!email) {
      return null;
    }

    if (email.startsWith('mailto:')) {
      email = email.slice('mailto:'.length);
    }

    if (!email.includes('@') || email.includes(' ')) {
      return null;
    }

    if (email.endsWith('@neu.edu')) {
      email = `${email.split('@')[0]}@northeastern.edu`;
    }

    return email.toLowerCase().trim();
  }


  static standardizePhone(phone) {
    if (!phone) {
      return null;
    }

    phone = phone.trim();

    if (phone.startsWith('tel:')) {
      phone = phone.slice('tel:'.length).trim();
    }

    let digitsOnly = phone.replace(/[^0-9]/gi, '');


    if (phone.startsWith('+1') && digitsOnly.length === 11) {
      digitsOnly = digitsOnly.slice(1);
    }

    if (digitsOnly.length !== 10) {
      return null;
    }

    return digitsOnly;
  }

  // Parses the google scholar id from a link that should contain a google scholar link.
  // Get the Google Scholar ID with this: https://scholar.google.com/citations?user=[id here]
  static parseGoogleScolarLink(link) {
    if (!link) {
      return null;
    }

    const userId = new URI(link).query(true).user;
    if (!userId && link) {
      Macros.log('Error parsing google url', link);
      return null;
    }
    return userId;
  }


  // Gets the base hostname from a url.
  // fafjl.google.com -> google.com
  // subdomain.bob.co -> bob.co
  // bob.co -> bob.co
  // This could be improved by using public lists of top-level domains.
  static getBaseHost(url) {
    const homepage = new URI(url).hostname();
    if (!homepage || homepage === '') {
      Macros.error('could not find homepage of', url);
      return null;
    }

    const match = homepage.match(/[^.]+\.[^.]+$/i);
    if (!match) {
      Macros.error('homepage match failed...', homepage);
      return null;
    }
    return match[0];
  }


  // http://stackoverflow.com/questions/4009756/how-to-count-string-occurrence-in-string/7924240#7924240
  static occurrences(string, subString, allowOverlapping) {
    string += '';
    subString += '';
    if (subString.length <= 0) {
      return (string.length + 1);
    }

    let n = 0;
    let pos = 0;
    const step = allowOverlapping ? 1 : subString.length;

    while (true) {
      pos = string.indexOf(subString, pos);
      if (pos >= 0) {
        ++n;
        pos += step;
      } else {
        break;
      }
    }
    return n;
  }

  static async getAllEnvVariables() {
    let configFileName = '/etc/searchneu/config.json';

    let exists = await fs.exists(configFileName);

    // Also check /mnt/c/etc... in case we are running inside WSL.
    if (!exists) {
      configFileName = '/mnt/c/etc/searchneu/config.json';
      exists = await fs.exists(configFileName);
    }

    if (!exists) {
      return {};
    }

    const body = await fs.readFile(configFileName);

    return JSON.parse(body);
  }

  static async getEnvVariable(name) {
    if (!envVariablesPromise) {
      envVariablesPromise = this.getAllEnvVariables();
    }

    return (await envVariablesPromise)[name];
  }

  // Log an event to amplitude. Same function signature as the function for the frontend.
  static async logAmplitudeEvent(type, event) {
    if (!Macros.PROD) {
      return null;
    }

    const data = {
      event_type: type,
      device_id: 'backend',
      session_id: Date.now(),
      event_properties: event,
    };

    return amplitude.track(data);
  }


  // This is for programming errors. This will cause the program to exit anywhere.
  // This *should* never be called.
  static critical(...args) {
    Macros.error(...args);
    process.exit(1);
  }

  // Use this for stuff that should never happen, but does not mean the program cannot continue.
  // This will continue running in dev, but will exit on CI
  // Will log stack trace
  // and cause CI to fail
  // so CI will send an email
  static async error(...args) {
    super.error(...args);

    if (Macros.PROD) {
      // If running on Travis, just exit 1 and travis will send off an email.
      if (process.env.CI) {
        process.exit(1);

      // If running on AWS, tell rollbar about the error so rollbar sends off an email.
      } else {
        const rollbarKey = await Macros.getEnvVariable('rollbarPostServerItemToken');
        rollbar.init(rollbarKey);

        const stack = (new Error()).stack;

        // The middle object can include any properties and values, much like amplitude.
        args.stack = stack;

        if (args.length === 0) {
          args.push('Error had no message?');
        }

        if (args[0] instanceof Error) {
          // The middle object can include any properties and values, much like amplitude.
          rollbar.handleError(args[0], args, () => {
            // And kill the process to recover.
            // forver.js will restart it.
            process.exit(1);
          });
        } else {
          rollbar.error(args[0], args, () => {
            process.exit(1);
          });
        }
      }
    }
  }

  // Use console.warn to log stuff during testing

  static verbose(...args) {
    if (!process.env.VERBOSE) {
      return;
    }

    console.log(...args);
  }
}


Macros.PUBLIC_DIR = path.join('public', 'data');
Macros.DEV_DATA_DIR = path.join('cache', 'dev_data');

// For iterating over every letter in a couple different places in the code.
Macros.ALPHABET = 'maqwertyuiopsdfghjklzxcvbn';


Macros.verbose('Starting in verbose mode.');


async function handleUncaught(err) {
  console.log('Error: An unhandledRejection occurred.');
  console.log(`Rejection Stack Trace: ${err.stack}`);
  Macros.error(err.stack);
}


// Sometimes it helps debugging to enable this test mode too.
if ((Macros.PROD || Macros.DEV || 1) && !global.addedRejectionHandler) {
  global.addedRejectionHandler = true;
  process.on('unhandledRejection', handleUncaught);
  process.on('uncaughtException', handleUncaught);
}


export default Macros;
