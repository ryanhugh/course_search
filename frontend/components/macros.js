/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import isMobile from 'is-mobile';
import ReactTooltip from 'react-tooltip';
import commonMacros from '../../common/abstractMacros';


// Macros and other utility constants and things that are just relevant in the frontend.


// Used for debounceTooltipRebuild
let tooltipTimer = null;

class Macros extends commonMacros {
  // Text area used for the copy method, below
  static copyTextArea = null;

  static log(...args) {
    // Don't log stuff in prod mode
    if (Macros.PROD) {
      return;
    }

    commonMacros.log(...args);
  }

  // Call this to reload the tooltip rendering on the entire page.
  // Feel free to call as often as you want, this has internal debouncing so it will only rebuild the tooltips 20ms after the last update call.
  // Currently used in just ClassPanel.js.
  static debounceTooltipRebuild() {
    clearTimeout(tooltipTimer);
    tooltipTimer = setTimeout(ReactTooltip.rebuild.bind(ReactTooltip), 20);
  }

  // Log an event to amplitude. Same function signature as the function for the backend/.
  // This call just uses the script included in index.js
  static async logAmplitudeEvent(type, event) {
    if (!Macros.PROD) {
      return;
    }

    if (!window.amplitude) {
      Macros.error("Can't log to amplitude without amplitude script!");
      return;
    }
    window.amplitude.logEvent(type, event, (statusCode) => {
      if (statusCode !== 200) {
        Macros.log('Amplitude logging failed', statusCode);
      }
    });
  }

  // Takes in text to be copied to the clipboard.
  // This process is longer than it seems like it should be... (thanks JS)
  // There is a new APi
  // This code is gutted from the core of this module: https://github.com/zenorocha/clipboard.js#readme
  static copyToClipboard(input) {
    // Try to copy with the new API, if it exists
    if (navigator.clipboard && navigator.clipboard.writeText) {
      navigator.clipboard.writeText(input);
      return;
    }

    // If not, use a much longer process...
    const isRTL = document.documentElement.getAttribute('dir') === 'rtl';

    if (!this.copyTextArea) {
      this.copyTextArea = document.createElement('textarea');
    }

    this.copyTextArea.style.display = '';


    // Prevent zooming on iOS
    this.copyTextArea.style.fontSize = '12pt';
    // Reset box model
    this.copyTextArea.style.border = '0';
    this.copyTextArea.style.padding = '0';
    this.copyTextArea.style.margin = '0';
    // Move element out of screen horizontally
    this.copyTextArea.style.position = 'absolute';
    this.copyTextArea.style[isRTL ? 'right' : 'left'] = '-9999px';
    // Move element to the same position vertically
    const yPosition = window.pageYOffset || document.documentElement.scrollTop;
    this.copyTextArea.style.top = `${yPosition}px`;

    this.copyTextArea.setAttribute('readonly', '');
    this.copyTextArea.value = input;

    if (!document.body.contains(this.copyTextArea)) {
      document.body.appendChild(this.copyTextArea);
    }

    this.copyTextArea.select();
    this.copyTextArea.setSelectionRange(0, this.copyTextArea.value.length);

    try {
      document.execCommand('copy');
    } catch (err) {
      this.error('Cannot copy', err);
    }

    this.copyTextArea.style.display = 'none';
  }
}

// The backtick on the third row and all the backslashes need to be escaped.
// This was generated with this site http://patorjk.com/software/taag/#p=display&f=Stacey&t=Search%20NEU
const searchneu = `
   ____                     _       _   _ _____ _   _
 / ___|  ___  __ _ _ __ ___| |__   | \\ | | ____| | | |
 \\___ \\ / _ \\/ _\` | '__/ __| '_ \\  |  \\| |  _| | | | |
  ___) |  __/ (_| | | | (__| | | | | |\\  | |___| |_| |
 |____/ \\___|\\__,_|_|  \\___|_| |_| |_| \\_|_____|\\___/



 Hi There!

 We're looking for talented individuals who want to build great products
 that impact thousands of student's lives. Interested? Help us build Search NEU!

 Contact us at hey@searchneu.com :)
 `;

if (Macros.PROD) {
  commonMacros.log(searchneu);
}


// How many sections to show in a class panel by default.
Macros.sectionsShownByDefault = 3;

// How many sections to add when the user clicks the show more button.
Macros.sectionsAddedWhenShowMoreClicked = 5;

// If this number of section is shown, the show more button will just show the rest of them instead of showing just a couple more.
Macros.sectionsShowAllThreshold = 15;

Macros.searchEvent = 'customSearch';

// True if is a Phone or other mobile device (iPod). Will be false for iPads.
Macros.isMobile = isMobile();


// Definition of some enums for our prerequsities
Macros.prereqTypes = Object.freeze({
  PREREQ: Symbol('prereq'),
  COREQ: Symbol('coreq'),
  PREREQ_FOR: Symbol('prereqFor'),
  OPT_PREREQ_FOR: Symbol('optPrereqFor'),
});

export default Macros;
