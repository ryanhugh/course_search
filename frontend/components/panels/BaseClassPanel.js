/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import React from 'react';
import PropTypes from 'prop-types';

import RequisiteBranch from '../../../common/classModels/RequisiteBranch';
import Keys from '../../../common/Keys';
import macros from '../macros';

class BaseClassPanel extends React.Component {
  static propTypes = {
    aClass: PropTypes.object.isRequired,
  };

  constructor(props) {
    super(props);

    let sectionsShownByDefault;
    if (this.constructor.sectionsShownByDefault) {
      sectionsShownByDefault = this.constructor.sectionsShownByDefault;
    } else {
      sectionsShownByDefault = macros.sectionsShownByDefault;
    }

    // If this is desktop and there is exactly one section hidden by the button, just show them all.
    if (!macros.isMobile && this.props.aClass.sections.length === sectionsShownByDefault + 1) {
      sectionsShownByDefault++;
    }

    this.state = {
      prereqsPage: 0,
      coreqsPage: 0,
      prereqsForPage: 0,
      optPrereqsForPage: 0,
      renderedSections: this.props.aClass.sections.slice(0, sectionsShownByDefault),
      unrenderedSections: this.props.aClass.sections.slice(sectionsShownByDefault),
    };

    this.onShowMoreClick = this.onShowMoreClick.bind(this);
  }

  onShowMoreClick() {
    macros.log('Adding more sections to the bottom.');

    // Get the length of the our sections
    const rendered = this.state.renderedSections;
    const unrendered = this.state.unrenderedSections;

    const showAmount = (rendered.length > macros.sectionsShowAllThreshold) ?
      unrendered.length : macros.sectionsAddedWhenShowMoreClicked;

    this.setState({
      renderedSections: rendered.concat(unrendered.slice(0, showAmount)),
      unrenderedSections: unrendered.slice(showAmount, unrendered.length),
    });
  }

  // Prevents page reload and fires off new search without reloading.
  onReqClick(reqType, childBranch, event) {
    // Create the React element and add it to retVal
    const searchEvent = new CustomEvent(macros.searchEvent, { detail: `${childBranch.subject} ${childBranch.classId}` });
    window.dispatchEvent(searchEvent);
    event.preventDefault();

    // Rest of this function is analytics
    const classCode = `${childBranch.subject} ${childBranch.classId}`;
    let reqTypeString;

    switch (reqType) {
      case macros.prereqTypes.PREREQ:
        reqTypeString = 'Prerequisite';
        break;
      case macros.prereqTypes.COREQ:
        reqTypeString = 'Corequisite';
        break;
      case macros.prereqTypes.PREREQ_FOR:
        reqTypeString = 'Required Prerequisite For';
        break;
      case macros.prereqTypes.OPT_PREREQ_FOR:
        reqTypeString = 'Optional Prerequisite For';
        break;
      default:
        macros.error('unknown type.', reqType);
    }

    macros.logAmplitudeEvent('Requisite Click', {
      type: reqTypeString,
      subject: childBranch.subject,
      classId: childBranch.classId,
      classCode: classCode,
    });
  }

  // Render the Show More.. Button
  // This is the same on both desktop and mobile.
  getMoreSectionsButton() {
    if (this.state.unrenderedSections.length <= 0) {
      return null;
    }

    return (
      <div
        className='more-sections-button'
        role='button'
        tabIndex={ 0 }
        onClick={ this.onShowMoreClick }
      >
        Show More...
      </div>
    );
  }

  getCreditsString() {
    // Figure out the credits string
    let retStr = '';
    if (this.props.aClass.maxCredits !== this.props.aClass.minCredits) {
      retStr = `${this.props.aClass.minCredits} to `;
    }
    return `${retStr}${this.props.aClass.maxCredits} credits`;
  }

  // The argument wrapper func is optional
  // If it exists, it is called on when formatting the classes
  // It is called with a class
  // and can return either a string or a react element.
  getReqsString(reqType, aClass = this.props.aClass) {
    const retVal = [];

    // Keep track of which subject+classId combonations have been used so far.
    // If you encounter the same subject+classId combo in the same loop, skip the second one.
    // This is because there is no need to show (eg. CS 2500 and CS 2500 (hon)) in the same group
    // because only the subject and the classId are going to be shown.
    const processedSubjectClassIds = {};

    let childNodes;

    if (reqType === macros.prereqTypes.PREREQ) {
      childNodes = aClass.prereqs;
    } else if (reqType === macros.prereqTypes.COREQ) {
      childNodes = aClass.coreqs;
    } else if (reqType === macros.prereqTypes.PREREQ_FOR) {
      childNodes = aClass.prereqsFor;
    } else if (reqType === macros.prereqTypes.OPT_PREREQ_FOR) {
      childNodes = aClass.optPrereqsFor;
    } else {
      macros.error('Invalid prereqType', reqType);
    }

    childNodes.values.forEach((childBranch) => {
      // If the childBranch is a class
      if (!(childBranch instanceof RequisiteBranch)) {
        if (childBranch.isString) {
          // Skip if already seen
          if (processedSubjectClassIds[childBranch.desc]) {
            return;
          }
          processedSubjectClassIds[childBranch.desc] = true;

          retVal.push(childBranch.desc);
        } else {
          // Skip if already seen
          if (processedSubjectClassIds[childBranch.subject + childBranch.classId]) {
            return;
          }
          processedSubjectClassIds[childBranch.subject + childBranch.classId] = true;

          // When adding support for right click-> open in new tab, we might also be able to fix the jsx-a11y/anchor-is-valid errors.
          // They are disabled for now.
          const hash = `${childBranch.subject} ${childBranch.classId}`;

          const element = (
            <a
              key={ hash }
              href={ hash }
              tabIndex={ 0 }
              onClick={ (event) => { this.onReqClick(reqType, childBranch, event); } }
              className='reqClassLink'
            >
              {`${childBranch.subject} ${childBranch.classId}`}
            </a>
          );

          retVal.push(element);
        }
      } else if (reqType === macros.prereqTypes.PREREQ) {
        // Figure out how many unique classIds there are in the prereqs.
        const allClassIds = {};
        for (const node of childBranch.prereqs.values) {
          allClassIds[node.classId] = true;
        }

        // If there is only 1 prereq with a unique classId, don't show the parens.
        if (Object.keys(allClassIds).length === 1) {
          retVal.push(this.getReqsString(macros.prereqTypes.PREREQ, childBranch));
        } else {
          retVal.push(['(', this.getReqsString(macros.prereqTypes.PREREQ, childBranch), ')']);
        }
      } else {
        macros.error('Branch found and parsing coreqs?', childBranch);
      }
    });


    // Now insert the type divider ("and" vs "or") between the elements.
    // If we're parsing prereqsFor, we should use just a comma as a separator.
    // Can't use the join in case the objects are react elements
    if (reqType === macros.prereqTypes.PREREQ_FOR || reqType === macros.prereqTypes.OPT_PREREQ_FOR) {
      for (let i = retVal.length - 1; i >= 1; i--) {
        retVal.splice(i, 0, ', ');
      }
    } else {
      let type;
      if (reqType === macros.prereqTypes.PREREQ) {
        type = aClass.prereqs.type;
      } else if (reqType === macros.prereqTypes.COREQ) {
        type = aClass.coreqs.type;
      }
      for (let i = retVal.length - 1; i >= 1; i--) {
        retVal.splice(i, 0, ` ${type} `);
      }
    }

    if (retVal.length === 0) {
      return <span className='hint-text'>None</span>;
    }

    // retVal = retVal.join(' ' + this.prereqs.type + ' ')

    return retVal;
  }

  /**
   * Returns the 'page' of the specified prerequisite.
   *
   * @param {prereqTypes} prereqType type of prerequisite.
   */
  getStateValue(prereqType) {
    switch (prereqType) {
      case macros.prereqTypes.PREREQ:
        return this.state.prereqsPage;
      case macros.prereqTypes.COREQ:
        return this.state.coreqsPage;
      case macros.prereqTypes.PREREQ_FOR:
        return this.state.prereqsForPage;
      case macros.prereqTypes.OPT_PREREQ_FOR:
        return this.state.optPrereqsForPage;
      default:
        return -1;
    }
  }

  /**
   * Returns how many elements we should return from our array of prerequisites.
   * Note that we mutliply our value by two because every other value is ', '
   * We could probably make this better...
   *
   * @param {prereqTypes} prereqType type of prerequisite.
   */
  getShowAmount(prereqType) {
    const classesShownByDefault = 5;
    const stateValue = this.getStateValue(prereqType);
    return 2 * classesShownByDefault * (stateValue + 1);
  }

  /**
   * Returns the array that we should be displaying
   *
   * @param {prereqTypes} prereqType type of prerequisite.
   */
  optionalDisplay(prereqType) {
    const data = this.getReqsString(prereqType, this.props.aClass);

    if (Array.isArray(data)) {
      if (this.getStateValue(prereqType) >= 3) {
        return data;
      }

      const showAmt = this.getShowAmount(prereqType);

      if (showAmt < data.length) {
        data.length = showAmt;
      }

      if (typeof data[data.length - 1] === 'string') {
        data.length -= 1;
      }
    }

    return data;
  }

  /**
   * Returns the 'Show More' button of the prereqType, if one is needed.
   * @param {prereqTypes} prereqType type of prerequisite.
   */
  showMore(prereqType) {
    const data = this.getReqsString(prereqType, this.props.aClass);

    if (!Array.isArray(data) ||
      this.getStateValue(prereqType) >= 3 ||
      this.getShowAmount(prereqType) >= data.length) {
      return null;
    }

    return (
      <div
        className='prereq-show-more hint-text'
        tabIndex={ 0 }
        role='button'
        onClick={ () => {
          this.setState((prevState) => {
            switch (prereqType) {
              case macros.prereqTypes.PREREQ:
                return { prereqsPage: prevState.prereqsPage + 1 };
              case macros.prereqTypes.COREQ:
                return { prereqsPage: prevState.coreqsPage + 1 };
              case macros.prereqTypes.PREREQ_FOR:
                return { prereqsForPage: prevState.prereqsForPage + 1 };
              case macros.prereqTypes.OPT_PREREQ_FOR:
                return { optPrereqsForPage: prevState.optPrereqsForPage + 1 };
              default:
                return macros.error(`invalid prereq type!: ${prereqType}`);
            }
          });
        } }
      >Show More
      </div>
    );
  }

  // Just used for testing. Not used anywhere in the frontend.
  render() {
    return (
      <div>
        <div>
          {this.state.renderedSections.map((aClass) => {
            return Keys.create(aClass).getHash();
          })}
        </div>
        <div>
          {this.state.unrenderedSections.map((aClass) => {
            return Keys.create(aClass).getHash();
          })}
        </div>
      </div>
    );
  }
}

export default BaseClassPanel;
