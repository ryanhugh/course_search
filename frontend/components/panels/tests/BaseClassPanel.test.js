/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import React from 'react';
import renderer from 'react-test-renderer';

import mockData from './mockData';
import BaseClassPanel from '../BaseClassPanel';
import macros from '../../macros';


it('should not render a show more button', () => {
  const component = renderer.create(<BaseClassPanel aClass={ mockData.cs0210 } />);

  const button = component.getInstance().getMoreSectionsButton();

  expect(button).toMatchSnapshot();
});


it('should render a button', () => {
  const component = renderer.create(<BaseClassPanel aClass={ mockData.cs1210 } />);

  const button = component.getInstance().getMoreSectionsButton();

  expect(button).toMatchSnapshot();
});


it('clicking the button will cause more sections to be rendered', () => {
  const component = renderer.create(<BaseClassPanel aClass={ mockData.cs1210 } />);

  expect(component.getInstance().state.renderedSections.length).toBe(3);
  expect(component.getInstance().state.unrenderedSections.length).toBe(2);

  component.getInstance().onShowMoreClick();

  expect(component.getInstance().state.renderedSections.length).toBe(5);
  expect(component.getInstance().state.unrenderedSections.length).toBe(0);
});


it('should display credits string', () => {
  const component = renderer.create(<BaseClassPanel aClass={ mockData.cs1210 } />);

  let instance = component.getInstance();

  expect(instance.getCreditsString()).toBe('1 credits');

  const cs0210 = renderer.create(<BaseClassPanel aClass={ mockData.cs0210 } />);

  instance = cs0210.getInstance();

  expect(instance.getCreditsString()).toBe('1 to 49 credits');
});


it('should render prereqs string', () => {
  const component = renderer.create(<BaseClassPanel aClass={ mockData.cs0210 } />);

  const instance = component.getInstance();

  const prereqs = instance.getReqsString(macros.prereqTypes.PREREQ);

  expect(prereqs).toMatchSnapshot();
});


it('should render coreqs string', () => {
  const component = renderer.create(<BaseClassPanel aClass={ mockData.cs0210 } />);

  const instance = component.getInstance();

  const coreqs = instance.getReqsString(macros.prereqTypes.COREQ);

  expect(coreqs).toMatchSnapshot();
});
