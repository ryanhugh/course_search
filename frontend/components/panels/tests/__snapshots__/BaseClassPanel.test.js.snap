// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not render a show more button 1`] = `null`;

exports[`should render a button 1`] = `
<div
  className="more-sections-button"
  onClick={[Function]}
  role="button"
  tabIndex={0}
>
  Show More...
</div>
`;

exports[`should render coreqs string 1`] = `
Array [
  <a
    className="reqClassLink"
    href="CHEM 3506"
    onClick={[Function]}
    tabIndex={0}
  >
    CHEM 3506
  </a>,
  " and ",
  <a
    className="reqClassLink"
    href="CHEM 3507"
    onClick={[Function]}
    tabIndex={0}
  >
    CHEM 3507
  </a>,
]
`;

exports[`should render prereqs string 1`] = `
Array [
  Array [
    "(",
    Array [
      <a
        className="reqClassLink"
        href="CHEM 3401"
        onClick={[Function]}
        tabIndex={0}
      >
        CHEM 3401
      </a>,
      " or ",
      <a
        className="reqClassLink"
        href="CHEM 3431"
        onClick={[Function]}
        tabIndex={0}
      >
        CHEM 3431
      </a>,
    ],
    ")",
  ],
  " and ",
  Array [
    "(",
    Array [
      <a
        className="reqClassLink"
        href="CHEM 2321"
        onClick={[Function]}
        tabIndex={0}
      >
        CHEM 2321
      </a>,
      " or ",
      <a
        className="reqClassLink"
        href="CHEM 2331"
        onClick={[Function]}
        tabIndex={0}
      >
        CHEM 2331
      </a>,
    ],
    ")",
  ],
  " and ",
  Array [
    "(",
    Array [
      <a
        className="reqClassLink"
        href="CHEM 2313"
        onClick={[Function]}
        tabIndex={0}
      >
        CHEM 2313
      </a>,
      " or ",
      <a
        className="reqClassLink"
        href="CHEM 2317"
        onClick={[Function]}
        tabIndex={0}
      >
        CHEM 2317
      </a>,
    ],
    ")",
  ],
]
`;
